### PR TITLE
Fix copyfile_eeglab type conflicts

### LINKED ISSUE #1122: write_raw_bids does not produce a valid SET file
### Description of the problem

The function write_raw_bids does not produce a valid .set file that can be read by EEGLAB. When importing a .set file that was generated by write_raw_bids in EEGLAB, we systematically get the following error:

Screenshot (23)

### Steps to reproduce

```Python
import mne
import os
from mne_bids import write_raw_bids, BIDSPath

file = r'C:\Users\Lae\Documents\test_data\DCC0003_444597_V01_FACE.set'
bids_directory = r'C:\Users\Lae\Documents\test_data'
raw = mne.io.read_raw_eeglab(input_fname=file, preload=False, verbose=True)

os.makedirs(bids_directory + os.path.sep, exist_ok=True)
bids_directory = bids_directory + os.path.sep
bids_root = bids_directory
bids_basename = BIDSPath(subject='001', task='test', root=bids_root, acquisition='eeg', run=None)
bids_basename.update(session='V1')
write_raw_bids(raw, bids_basename, allow_preload=False, overwrite=False, verbose=True)
```

### Expected results

A valid EEGLAB file

### Actual results

File can't be open, see above error.

### Additional information

Platform:         Windows-10-10.0.19041-SP0
Python:           3.8.5 (tags/v3.8.5:580fbb0, Jul 20 2020, 15:57:54) [MSC v.1924 64 bit (AMD64)]
Executable:       C:\Users\Lae\Documents\hbcd-eeg2bids\Scripts\python.exe
CPU:              Intel64 Family 6 Model 142 Stepping 9, GenuineIntel: 4 cores
Memory:           Unavailable (requires "psutil" package)
mne:              1.2.0
numpy:            1.20.3 {OpenBLAS 0.3.13.dev with 4 threads}
scipy:            1.9.2
matplotlib:       3.6.1 {backend=TkAgg}

sklearn:          1.1.2
numba:            0.56.2
nibabel:          Not found
nilearn:          Not found
dipy:             Not found
openmeeg:         Not found
cupy:             Not found
pandas:           1.5.0
pyvista:          Not found
pyvistaqt:        Not found
ipyvtklink:       Not found
vtk:              Not found
qtpy:             Not found
ipympl:           Not found
pyqtgraph:        Not found
pooch:            v1.6.0

mne_bids:         0.12
mne_nirs:         Not found
mne_features:     0.2.1
mne_qt_browser:   Not found
mne_connectivity: Not found
mne_icalabel:     Not found

### PR DESCRIPTION
**Pull request by laemtl**
PR Description
--------------
Fix #1122

Merge checklist
--------------
Maintainer, please confirm the following before merging.
If applicable:

- [ ] All comments are resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/main/doc/whats_new.rst) is updated
- [ ] New contributors have been added to [CITATION.cff](https://github.com/mne-tools/mne-bids/blob/main/CITATION.cff)
- [x] PR description includes phrase "closes <#issue-number>"

### CODE REVIEW

File: doc/whats_new.rst [hunk 1]
```diff
@@ -52,6 +52,7 @@ Detailed list of changes
 ^^^^^^^^^^^^
 
 - Amending a dataset now works in cases where the newly-written data contains additional participant properties (new columns in ``participants.tsv``) not found in the existing dataset, by `Richard Höchenberger`_ (:gh:`1113`)
+- Fixes copyfile_eeglab to prevent data int conversion and eeg_checkset failure, by `Laetitia Fesselier`_ (:gh:`1122`)
```
Comment (hoechenberger) [inline doc/whats_new.rst:55]:
```suggestion
- Fixes :func:`~mne_bids.copyfiles.copyfile_eeglab` to prevent data type conversion leading to an ``eeg_checkset`` failure when trying to load the file in EEGLAB, by `Laetitia Fesselier`_ (:gh:`1122`)
```

**Review by hoechenberger — APPROVED** [review]

File: mne_bids/copyfiles.py [hunk 2]
```diff
@@ -564,6 +564,15 @@ def copyfile_eeglab(src, dest):
...
         eeg['data'] = new_value
 
+        # Make sure all values are saved as float
+        # (see https://github.com/mne-tools/mne-bids/issues/1122)
+        eeg['nbchan'] = float(eeg['nbchan'])
+        eeg['pnts'] = float(eeg['pnts'])
+        eeg['trials'] = float(eeg['trials'])
+        eeg['srate'] = float(eeg['srate'])
+        eeg['xmin'] = float(eeg['xmin'])
+        eeg['xmax'] = float(eeg['xmax'])
```
Comment (jasmainak) [inline mne_bids/copyfiles.py:574]: I fear this is not a future-proof solution ... the only thing we should be modifying is the pointer to the filename and leaving everything else unchanged

**Comment by sappelhoff** [comment]
circleci was a bit weird:

image

never saw that before ... anyone else? Manually restarting the workflow seems to have helped

**Comment by adam2392** [comment]
> circleci was a bit weird:
> 
> image
> 
> never saw that before ... anyone else? Manually restarting the workflow seems to have helped

We saw it in mne-connectivity too: https://github.com/mne-tools/mne-connectivity/pull/129#issuecomment-1460448899

File: mne_bids/copyfiles.py (hunk 2 above)
Comment (laemtl) [reply mne_bids/copyfiles.py:574]: @jasmainak Very good point. Your comment made me think of another solution. Passing `mat_dtype=True` to `loadmat` also fixes the issue without the need to cast back to float. Since I believe this is a better solution, I will update this PR. Thanks!

File: mne_bids/copyfiles.py [hunk 3]
```diff
@@ -533,18 +533,19 @@ def copyfile_eeglab(src, dest):
...
     uint16_codec = None
     eeg = loadmat(file_name=src, simplify_cells=False,
-                  appendmat=False, uint16_codec=uint16_codec)
+                  appendmat=False, uint16_codec=uint16_codec, mat_dtype=True)
     oldstyle = False
     if 'EEG' in eeg:
         eeg = eeg['EEG']
         oldstyle = True
 
+    has_fdt_link = False
```
Comment (hoechenberger) [inline mne_bids/copyfiles.py:542]: I would suggest to revert this change, as it makes the code flow less obvious

File: mne_bids/copyfiles.py (hunk 2 above)
Comment (jasmainak) [reply mne_bids/copyfiles.py]: excellent ! Glad to have nudged in a more optimal direction ;)

File: mne_bids/copyfiles.py (hunk 3 above)
Comment (laemtl) [reply mne_bids/copyfiles.py:542]: In the current implementation, the is some potential cases where has_fdt_link will not be set and a KeyError thrown.
It happened to me during testing so that's how I tried to fixed that. I can revert that change still.

File: mne_bids/copyfiles.py [hunk 4]
```diff
@@ -533,18 +533,19 @@ def copyfile_eeglab(src, dest):
     # structure and potentially breaks re-reading of the file
     uint16_codec = None
     eeg = loadmat(file_name=src, simplify_cells=False,
-                  appendmat=False, uint16_codec=uint16_codec)
+                  appendmat=False, uint16_codec=uint16_codec, mat_dtype=True)
```
Comment (sappelhoff) [inline mne_bids/copyfiles.py:536]: oh this is a sufficient fix?! How nice would that be

Comment (laemtl) [reply mne_bids/copyfiles.py:536]: Yes, it solved the pb on my end.

**Comment by sappelhoff** [comment]
you need to add yourself here as well @laemtl 

https://github.com/mne-tools/mne-bids/blob/0b1a61d7660b35438b78898878ef0c8b51d8716b/doc/authors.rst?plain=1#L42

**Comment by sappelhoff** [comment]
and here, just behind Moritz, please:

https://github.com/mne-tools/mne-bids/blob/0b1a61d7660b35438b78898878ef0c8b51d8716b/CITATION.cff#L154

File: mne_bids/copyfiles.py (hunk 3 above)
Comment (sappelhoff) [reply mne_bids/copyfiles.py:542]: For me both ways are fine. Thanks for figuring out that there is one potential flow that previously lead to an error!

**Review by sappelhoff — APPROVED** [review]

**Comment by sappelhoff** [comment]
Thanks @laemtl 🚀
